drivers/serial/uart_xlnx_ps.c

```diff
@@ -313,6 +387,457 @@ static void uart_xlnx_ps_poll_out(struct device *dev, unsigned char c)
+ *
+ * It is assumed that the Modem Control Register contents that are
+ * being modified within this function come with the bits modified
+ * by this function already masked out by the caller.
+ *
+ * @param modemcr_reg Modem Control Register contents to which
+ *                    the flow control configuration shall be added.
+ * @param hwctrl Enumeration value to be converted to a bit mask.
+ *
+ * @return Modem Control Register contents to which the data bit
+ *         setting has been added.
+ */
+static inline u32_t uart_xlnx_ps_cfg2ll_hwctrl(
+	u32_t modemcr_reg,
+	enum uart_config_flow_control hwctrl)
+{
+	/*
+	 * Translate the new flow control configuration to the modem
+	 * control register's bit [5] (FCM):
+	 *  0b : no flow control
+	 *  1b : RTS/CTS
+	 */
+
+	if (hwctrl == UART_CFG_FLOW_CTRL_RTS_CTS) {
+		modemcr_reg |= XUARTPS_MODEMCR_FCM_RTS_CTS;
+	} else { /* effectively: UART_CFG_FLOW_CTRL_NONE */
+		modemcr_reg |= XUARTPS_MODEMCR_FCM_NONE;
+	}
+
+	return modemcr_reg;
+}
+
+/**
+ * @brief Configures the UART device at run-time.
+ *
+ * Configures the UART device at run-time according to the
+ * configuration data provided by the caller.
+ *
+ * @param dev UART device struct
+ * @param cfg The configuration parameters to be applied.
+ *
+ * @return 0 if the configuration completed successfully, ENOTSUP
+ *         error if an unsupported configuration parameter is detected.
+ */
+static int uart_xlnx_ps_configure(struct device *dev,
+	const struct uart_config *cfg)
+{
+	struct uart_xlnx_ps_dev_config *dev_cfg =
+	(struct uart_xlnx_ps_dev_config *)DEV_CFG(dev);
+
+	u32_t reg_base    = dev_cfg->uconf.regs;
+	u32_t mode_reg    = 0;
+	u32_t modemcr_reg = 0;
+
+	/* Controller supports only RTS/CTS flow control */
+	if (cfg->flow_ctrl != UART_CFG_FLOW_CTRL_NONE) {
+		if (cfg->flow_ctrl != UART_CFG_FLOW_CTRL_RTS_CTS) {
+			return -ENOTSUP;
+		}
+	}
+
+	/* Controller doesn't support 0.5 stop bits */
+	if (cfg->stop_bits == UART_CFG_STOP_BITS_0_5) {
+		return -ENOTSUP;
+	}
+
+	/* Controller doesn't support 5 or 9 data bits */
+	if (cfg->data_bits == UART_CFG_DATA_BITS_5
+		|| cfg->data_bits == UART_CFG_DATA_BITS_9) {
+		return -ENOTSUP;
+	}
+
+	/* Disable the controller before modifying any config registers */
+
+	xlnx_ps_disable_uart(reg_base);
+
+	/* Set the baud rate */
+
+	set_baudrate(dev, cfg->baudrate);
+	dev_cfg->baud_rate = cfg->baudrate;
+
+	/* Read the current mode register & modem control register values */
+
+	mode_reg    = sys_read32(reg_base + XUARTPS_MR_OFFSET);
+	modemcr_reg = sys_read32(reg_base + XUARTPS_MODEMCR_OFFSET);
+
+	/* Mask out all items that might be re-configured */
+
+	mode_reg    &= (~XUARTPS_MR_PARITY_MASK);
+	mode_reg    &= (~XUARTPS_MR_STOPMODE_MASK);
+	mode_reg    &= (~XUARTPS_MR_CHARLEN_MASK);
+	modemcr_reg &= (~XUARTPS_MODEMCR_FCM_MASK);
+
+	/* Assemble the updated registers */
+
+	mode_reg    = uart_xlnx_ps_cfg2ll_parity(mode_reg, cfg->parity);
```

Comment:
I wonder if it would be better to make `uart_xlnx_ps_cfg2ll_*` funcs take a pointer to the `mode_reg` and return a boolean value indicating whether that configuration is valid, instead of having validity checks at two different locations.

```
static inline bool uart_xlnx_ps_cfg2ll_parity(u32_t *mode_reg, enum uart_config_parity parity)
```

```
if (!uart_xlnx_ps_cfg2ll_parity(&mode_reg, cfg->parity)) {
	return -ENOTSUP;
}
```